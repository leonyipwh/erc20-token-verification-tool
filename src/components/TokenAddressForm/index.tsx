import React, { useState, useEffect, useContext } from 'react';
import { initProvider } from '../../helpers/infuraHelper';
import { networks } from '../../constants/networks';
import { ITokenForm } from '../../types/tokenVerification';
import TokenContext from '../../context/tokenFormContext';
import SnackbarContext from '../../context/snackbarContext';
import MenuItem from '@mui/material/MenuItem';
import TextField from '@mui/material/TextField';
import Grid from '@mui/material/Grid';
import Select, { SelectChangeEvent } from '@mui/material/Select';

const errorMsg = {
  invalidAddress: 'Invalid Contract Address',
  infuraProvider: 'Unable to connect Infura'
};

const TokenAddressForm = () => {
  const [error, setError] = useState<string | null>(null);
  const [formValue, setFormValue] = useState<ITokenForm>({
    network: networks[0],
    address: ''
  });
  const { setSubmitForm } = useContext(TokenContext);
  const { snackbarState, setSnackbarState } = useContext(SnackbarContext);

  const handleNetworkChange = (event: SelectChangeEvent) => {
    setError(null);
    const network = networks.find((n) => n.name === event.target.value);
    if (!network) return;

    setFormValue({
      network,
      address: ''
    });
    setSubmitForm({ ...formValue, address: '' });
  };

  const handleAddressChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setFormValue({ ...formValue, address: event.target.value });
  };

  const getProvider = async () => {
    try {
      const provider = await initProvider(formValue.network.name);
      setFormValue({ ...formValue, provider });
    } catch (err) {
      console.error(err);
      setSnackbarState({ ...snackbarState, open: true, text: errorMsg.infuraProvider });
    }
  };

  const setProviderAndSubmitForm = async () => {
    if (!formValue.address || !formValue.provider) return;
    try {
      const code = await formValue.provider.getCode(formValue.address);
      if (code === '0x') throw new Error(errorMsg.invalidAddress);
      setSubmitForm({ ...formValue, provider: formValue.provider });
    } catch (err) {
      if (err.status === 401) {
        setSnackbarState({ ...snackbarState, open: true, text: err.body });
      } else {
        setError(errorMsg.invalidAddress);
      }

      setSubmitForm({ ...formValue, address: '' });
    }
  };

  useEffect(() => {
    getProvider();
  }, [formValue.network]);

  useEffect(() => {
    const addressLength = formValue.address.length;
    if (
      (addressLength > 0 && addressLength !== 42) ||
      (addressLength === 0 && formValue.provider)
    ) {
      setError(errorMsg.invalidAddress);
      setSubmitForm({ ...formValue, address: '' });
      return;
    }
    setError(null);
    setProviderAndSubmitForm();
  }, [formValue.address]);

  return (
    <Grid
      container
      spacing={2}
      sx={{ mt: 0 }}
    >
      <Grid
        item
        xs={12}
        md={2}
      >
        <Select
          id="ethNetworkSelect"
          value={formValue.network.name}
          onChange={handleNetworkChange}
          sx={{ minWidth: '100%', textTransform: 'capitalize' }}
        >
          {networks.map((n, index) => (
            <MenuItem
              key={index}
              value={n.name}
              sx={{ textTransform: 'capitalize' }}
            >
              {n.name}
            </MenuItem>
          ))}
        </Select>
      </Grid>
      <Grid
        item
        xs={12}
        md={10}
      >
        <TextField
          id="contractAddress"
          label="Contract Address"
          value={formValue.address}
          error={!!error}
          helperText={error}
          sx={{ minWidth: '100%' }}
          onChange={handleAddressChange}
        />
      </Grid>
    </Grid>
  );
};

export default TokenAddressForm;
