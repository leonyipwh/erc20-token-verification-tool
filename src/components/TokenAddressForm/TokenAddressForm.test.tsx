import { render, waitFor } from '@testing-library/react';
import TokenAddressForm from './index';

describe('<TokenAddressForm>', () => {
  it('should render', async () => {
    const { asFragment } = render(<TokenAddressForm />);
    await waitFor(() => expect(asFragment()).toMatchSnapshot());
  });
});
