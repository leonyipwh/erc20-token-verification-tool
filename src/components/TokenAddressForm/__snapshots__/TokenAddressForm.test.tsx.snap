// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TokenAddressForm> should render 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1c2jgnc-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-2 css-12xeub0-MuiGrid-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary  css-whju9g-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="ethNetworkSelect"
          class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="ethNetworkSelect"
          role="button"
          tabindex="0"
        >
          mainnet
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="mainnet"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-hdw1oc"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-10 css-1yysv6i-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-ir90mj-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="contractAddress"
          id="contractAddress-label"
        >
          Contract Address
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="contractAddress"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Contract Address
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
